README.txt
redomino.breadcrumbsbrain
=========================

redomino.breadcrumbsbrain is a new plugin for Plone CMS that will make it much easier
to search for specific objects in a Plone-based portal or intranet by showing the 
breadcrumbs paths along with Titles and Descriptions in the search results.

We all know that common users do not provide objects with proper titles and descriptions 
(sometimes they just do not provide them at all), which makes it hard to figure out 
if what you found is actually what you were looking for. Then you open every link in
the search results to find it out, which may take you quite a time. What you actually need 
in a situation like this is CONTEXT.

And context is what you get!
redomino.breadcrumbsbrain displays the breadcrumbs of each object in the search results,
helping you in figuring out what those objects are about: are they from 2010 or 2011?
are they in the 'Notices' archive or in the 'News' archive? and so on.


BE CAREFUL!
This is not a standalone package. If you want you can call the breadcrumbs_renderer
view provided by this package (for example you can insert this code into the
search.pt template)::

    <tal:block tal:replace="structure result/@@breadcrumbs_renderer|nothing" />

This way you will be able to display where your search results items are placed.

Many thanks to AUSL Bologna.

Technical details
-----------------

- redomino.breadcrumbsbrain.interfaces.IBrain
  This is a marker interface applied to all brains

- @@breadcrumbs_view
  This is a browser view with a method named breadcrumbs.
  Once called it will return the breadcrumbs info for a 
  particular brain. No getObject is needed, just a
  catalog call for each brain.
  Note well: if you want to use this browser view intensively you should
  STRONGLY consider to add a cache policy (it is not provided by default).

  Data format returned::

      [
       {'Title': 'News'),
        'absolute_url': 'http://localhost:8080/Plone/news'
       },
       {'Title': '2011'),
        'absolute_url': 'http://localhost:8080/Plone/news/2011'
       },
      ]

- @@breadcrumbs_renderer
  This is a browser view that let you show the breadcrumbs info
  of a particular brain.

  Data format displayed::

      News -> 2011

Authors
-------

- Davide Moro <davide.moro AT redomino.com>



docs/HISTORY.txt
Changelog
=========

0.5 (unreleased)
----------------

- Nothing changed yet.


0.4 (2012-07-16)
----------------

- fix template for duplicates ids [sithmel]


0.3 (2012-06-01)
----------------

- Added i18n
  [giacomos]


0.2 (2012-05-03)
----------------

- Added .gitignore file
  [davidemoro]

- Modified breadcrumbs brain format (added "Home ->")
  [davidemoro]


0.1 (2012-04-02)
----------------

- Initial release
  [davidemoro]


redomino/breadcrumbsbrain/tests.py
# Copyright (c) 2010 Redomino srl (http://redomino.com)
# Authors: Davide Moro <davide.moro@redomino.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.


import unittest

#from zope.testing import doctestunit
#from zope.component import testing
from Testing import ZopeTestCase as ztc

from Products.Five import fiveconfigure
from Products.PloneTestCase import PloneTestCase as ptc
from Products.PloneTestCase.layer import PloneSite
ptc.setupPloneSite()

import redomino.breadcrumbsbrain


class TestCase(ptc.PloneTestCase):

    class layer(PloneSite):

        @classmethod
        def setUp(cls):
            fiveconfigure.debug_mode = True
            ztc.installPackage(redomino.breadcrumbsbrain)
            fiveconfigure.debug_mode = False

        @classmethod
        def tearDown(cls):
            pass


def test_suite():
    return unittest.TestSuite([

        # Unit tests
        #doctestunit.DocFileSuite(
        #    'README.txt', package='redomino.breadcrumbsbrain',
        #    setUp=testing.setUp, tearDown=testing.tearDown),

        #doctestunit.DocTestSuite(
        #    module='redomino.breadcrumbsbrain.mymodule',
        #    setUp=testing.setUp, tearDown=testing.tearDown),


        # Integration tests that use PloneTestCase
        #ztc.ZopeDocFileSuite(
        #    'README.txt', package='redomino.breadcrumbsbrain',
        #    test_class=TestCase),

        #ztc.FunctionalDocFileSuite(
        #    'browser.txt', package='redomino.breadcrumbsbrain',
        #    test_class=TestCase),

        ])

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


redomino/breadcrumbsbrain/interfaces.py
# Copyright (c) 2010 Redomino srl (http://redomino.com)
# Authors: Davide Moro <davide.moro@redomino.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.


from zope.interface import Interface

class IBrain(Interface):
     """ Marker interface
     """



redomino/breadcrumbsbrain/breadcrumbs.py
# Copyright (c) 2010 Redomino srl (http://redomino.com)
# Authors: Davide Moro <davide.moro@redomino.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.

from Acquisition import aq_inner
from zope.interface import implements
from zope.component import getMultiAdapter

from Products.Five import BrowserView
from Products.Five.browser.pagetemplatefile import ViewPageTemplateFile
from Products.CMFPlone.browser.interfaces import INavigationBreadcrumbs
from Products.CMFPlone.browser.navigation import get_view_url
from Products.CMFPlone.browser.navigation import utils

class BrainBreadcrumbs(BrowserView):
    """ This view let you get a breadcrumbs starting from a brain, no getObject
        is needed.
        As result it returns a list of dictionaries,

        The overhead is proportional to the batch site (number of elements for each page).
        You can reduce the batch size if needed or implement some memoized policy.
    """
    implements(INavigationBreadcrumbs)

    def breadcrumbs(self):

        context = aq_inner(self.context)
        request = self.request
        ct = getMultiAdapter((context, request), name=u'plone_tools').catalog()
        query = {}

        currentPath = context.getPath()

        query['path'] = {'query':currentPath, 'navtree':1, 'depth': 0}
        rawresult = ct(**query)
        dec_result = [(len(r.getPath()),r) for r in rawresult]
        dec_result.sort()

        portal_state = getMultiAdapter((context, request), name=u'plone_portal_state')
        rootPath = portal_state.navigation_root_path()

        # Build result dict
        result = []
        for r_tuple in dec_result:
            item = r_tuple[1]

            # Don't include it if it would be above the navigation root
            itemPath = item.getPath()
            if rootPath.startswith(itemPath):
                continue

            id, item_url = get_view_url(item)
            data = {'Title': utils.pretty_title_or_id(context, item),
                    'absolute_url': item_url}
            result.append(data)
        return result

class BrainBreadcrumbsRenderer(BrowserView):
    """
    Breadcrumbs renderer view
    """
    __call__ = ViewPageTemplateFile('templates/breadcrumbs_renderer.pt')

